File: LieInBed-App/src/main/java/io/github/ctimet/lieinbedapp/gui/ui/WarnList.java
package io.github.ctimet.lieinbedapp.gui.ui;

import io.github.ctimet.lieinbedapp.gui.animation.AnimationUtil;
import io.github.ctimet.lieinbedapp.gui.util.CSSStyle;
import io.github.ctimet.lieinbedapp.gui.util.SVGUtil;
import io.github.ctimet.lieinbedapp.gui.util.Task;
import javafx.beans.binding.Bindings;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.geometry.Insets;
import javafx.scene.Node;
import javafx.scene.control.Label;
import javafx.scene.layout.VBox;
import javafx.scene.paint.Color;
import kotlin.jvm.internal.Intrinsics;
import org.jetbrains.annotations.NotNull;

public class WarnList extends VBox {
    private final ObservableList<Node> content = FXCollections.observableArrayList();
    private final int space;
    private final int layoutX;

    public WarnList(int eachSpace, int x, int ly) {
        this.space = eachSpace;
        this.layoutX = x;
        this.setLayoutY(ly);
    }

    public final void showWarn(@NotNull String warn) {
        Intrinsics.checkNotNullParameter(warn, "warn");
        final VBox box = new VBox();
        Label title = new Label("Warning");
        title.setGraphic(SVGUtil.warning(Bindings.createObjectBinding(() -> Color.RED), 1.0D, 1.0D));
        title.setTextFill(Color.RED);
        Label message = new Label(warn);
        box.getChildren().addAll(title, message);
        CSSStyle.addDefaultStyle(box, "vBox");
        this.setSpacing(this.space);
        box.setSpacing(10.0D);
        box.setPadding(new Insets(5.0D, 10.0D, 5.0D, 10.0D));
        box.setPrefWidth(207.0D);
        this.content.add(box);
        this.getChildren().add(box);
        AnimationUtil.takeXTranslateAnimation(box, 1085.0D, this.layoutX, 300);
        Task.delayRunInUIThread(() -> {
            AnimationUtil.takeXTranslateAnimation(box, box.getBoundsInParent().getMinX(), 1085.0D, 180);
            Task.delayRunInUIThread(() -> WarnList.this.getChildren().remove(box), 320);
        }, 4000);
    }
}

File: LieInBed-App/src/main/kotlin/io/github/ctimet/lieinbedapp0/gui/ui/WarnList.kt
package io.github.ctimet.lieinbedapp0.gui.ui

import io.github.ctimet.lieinbedapp0.gui.animation.AnimationUtil
import io.github.ctimet.lieinbedapp0.gui.util.CSSStyle
import io.github.ctimet.lieinbedapp0.gui.util.SVGUtil
import io.github.ctimet.lieinbedapp0.task.Task
import javafx.beans.binding.Bindings
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.geometry.Insets
import javafx.scene.control.Label
import javafx.scene.layout.VBox
import javafx.scene.paint.Color
import javafx.scene.paint.Paint

class WarnList(eachSpace: Int, x: Int, ly: Int) : VBox() {
    private val content: ObservableList<VBox> = FXCollections.observableArrayList()
    private val space = eachSpace
    private val layoutX = x

    init {
        layoutY = ly.toDouble()
    }

    fun showWarn(warn: String) {
        val box = VBox()
        val title = Label("Warning")
        title.graphic = SVGUtil.warning(
            Bindings.createObjectBinding(fun (): Paint {
                return Color.RED
            }), 1.0, 1.0
        )
        title.textFill = Color.RED
        val message = Label(warn)
        box.children.addAll(title, message)

        CSSStyle.addDefaultStyle(box, "vBox")
        spacing = space.toDouble()
        box.spacing = 10.0
        box.padding = Insets(5.0, 10.0, 5.0, 10.0)
        box.prefWidth = 207.0

        content.add(box)
        children.add(box)
        //执行动画
        AnimationUtil.takeXTranslateAnimation(box, 1085.0, layoutX.toDouble(), 300)
        Task.delayRunInUIThread({
            AnimationUtil.takeXTranslateAnimation(box, box.boundsInParent.minX, 1085.0, 180)
            Task.delayRunInUIThread({
                children.remove(box)
            }, 320)
        }, 4000)
    }
}